src/fs/data_compaction.cc

```diff
@@ -0,0 +1,397 @@
+            });
+        }).handle_exception([this](std::exception_ptr ptr) {
+            for (auto& cluster_id : _compacted_cluster_ids) {
+                _metadata_log.finish_writing_data_cluster(cluster_id);
+            }
+            mlogger.warn("Exception while flushing log after compaction: cannot free compacted clusters.");
+            return make_exception_future(std::move(ptr));
+        });
+    });
+}
+
+future<> data_compaction::read_data_vectors(std::vector<compacted_data_vec> data_vecs) {
+    // TODO: We could read the whole cluster when we read more data from it than some threshold (for example
+    //       more than 40%).
+    return do_with(device_reader(_metadata_log._device, _metadata_log._alignment), std::move(data_vecs),
+            [this](device_reader& reader, std::vector<compacted_data_vec>& data_vecs) {
+        return do_for_each(data_vecs, [&reader](compacted_data_vec& vec) {
+            return reader.read(vec._data.get_write(), vec._prev_disk_offset, vec._data.size()).then(
+                    [expected_read_len = vec._data.size()](size_t read_len) {
+                if (read_len != expected_read_len) {
+                    return make_exception_future<>(input_output_exception());
+                }
+                return now();
+            });
+        }).then([this, &data_vecs] {
+            return group_data_into_clusters(std::move(data_vecs));
+        });
+    });
+}
+
+future<> data_compaction::group_data_into_clusters(std::vector<compacted_data_vec> read_data_vecs) {
+    _metadata_log.throw_if_read_only_fs();
+
+    // TODO: we could group data vecs by file and merge them to perform some defragmentation
+
+    // First divide data into clusters and calculate data offsets in new clusters
+    std::vector<std::vector<compacted_data_vec>> grouped_data_vecs;
+    std::vector<compacted_data_vec> memory_data_vecs;
+    {
+        // TODO: We could try to avoid splitting the data using smarter strategies. That problem is similar to the multiple
+        // knapsack problem where knapsacks are clusters and items are data vectors. In that case we would need
+        // to remember that the main purpose of the compaction isn't smart placement of data vectors and defragmentation
+        // but releasing clusters.
+        bool should_add_new_group = true;
+        size_t in_cluster_offset = 0;
+        for (auto& file_data_vec : read_data_vecs) {
+            size_t remaining_data_size = file_data_vec._data.size();
+            size_t data_offset = 0;
+            while (in_cluster_offset + remaining_data_size > _metadata_log._cluster_size) {
+                if (should_add_new_group) {
+                    grouped_data_vecs.emplace_back(std::vector<compacted_data_vec> {});
+                    should_add_new_group = false;
+                }
+
+                size_t part_size = _metadata_log._cluster_size - in_cluster_offset;
+                grouped_data_vecs.back().emplace_back(compacted_data_vec {
```

Comment:
I am not sure about style, but semantically it should be the same.
```suggestion
                grouped_data_vecs.back().push_back({
```